core2/pkg/accounting/accounting_internal.go

```diff
@@ -390,6 +391,125 @@ func internalCommitAllocation(b *internalBucket, allocId accAllocId) {
 	b.Mu.Unlock()
 }
 
+func internalUpdateAllocation(actor rpc.Actor, now time.Time, b *internalBucket, allocationId accAllocId, newQuota util.Option[int64], newStart util.Option[fndapi.Timestamp], newEnd util.Option[fndapi.Timestamp]) *util.HttpError {
+	b.Mu.Lock()
+	defer b.Mu.Unlock()
+
+	actor.Membership[actor.Project.Value].Satisfies(rpc.ProjectRoleAdmin)
+	reference := actor.Username
+	if actor.Project.Present {
+		reference = string(actor.Project.Value)
+	}
+	iOwnerByActor := internalOwnerByReference(reference)
+
+	if iOwnerByActor == nil {
+		return util.HttpErr(http.StatusNotFound, "Unknown owner")
+	}
+	iAlloc, ok := b.AllocationsById[allocationId]
+	if !ok {
+		return util.HttpErr(http.StatusNotFound, "Unknown allocation")
+	}
+	iWallet, ok := b.WalletsById[iAlloc.BelongsTo]
+	if !ok {
+		return util.HttpErr(http.StatusNotFound, "Unknown wallet (bad internal state?)")
+	}
+	iAllocGroup, ok := iWallet.AllocationsByParent[iAlloc.Parent]
+	if !ok {
+		return util.HttpErr(http.StatusNotFound, "Unknown allocation group (bad internal state?)")
+	}
+
+	// Check if you have granted the allocation
+	iParent, ok := b.WalletsById[iAlloc.Parent]
+	if !ok {
+		return util.HttpErr(http.StatusNotFound, "Parent unknown")
+	}
+	if iParent.OwnedBy != iOwnerByActor.Id {
+		return util.HttpErr(http.StatusForbidden, "You are not allowed to modify this allocation")
+	}
+
+	if iAlloc.Retired {
+		return util.HttpErr(http.StatusForbidden, "You cannot update a retired allocation, it has already expired!")
+	}
+	if iAlloc.Start.Before(now) && newStart.Present {
+		return util.HttpErr(http.StatusForbidden, "You cannot change the starting time of an allocation which has already started")
+	}
+	proposedNewStart := iAlloc.Start
+	if newStart.Present {
+		proposedNewStart = newStart.Value.Time()
+	}
+	proposedNewEnd := iAlloc.End
+	if newEnd.Present {
+		proposedNewEnd = newEnd.Value.Time()
+	}
+	if proposedNewStart.After(proposedNewEnd) {
+		return util.HttpErr(http.StatusForbidden, "This update would make the allocation invalid. An allocation cannot start after it has ended.")
+	}
+	if newQuota.Present && newQuota.Value < 0 {
+		errorMessage := fmt.Sprintf("You cannot set a negative quota for an allocation (%d)", newQuota.Value)
+		return util.HttpErr(http.StatusForbidden, errorMessage)
+	}
+
+	proposedNewQuota := iAlloc.Quota
+	if newQuota.Present {
+		proposedNewQuota = newQuota.Value
+		delta := newQuota.Value - iAlloc.Quota
+		activeQuota := lInternalGroupTotalQuotaFromActiveAllocations(b, iAllocGroup)
+		activeUsage := iAllocGroup.TreeUsage
+
+		if activeQuota+delta < activeUsage {
+			return util.HttpErr(http.StatusForbidden, "You cannot decrease the quota below the current usage!")
+		}
+	}
+
+	iAlloc.Dirty = true
+	iAlloc.Start = proposedNewStart
+	iAlloc.End = proposedNewEnd
+	iAlloc.Quota = proposedNewQuota
+
+	lInternalMarkSignificantUpdate(b, now, iWallet)
+	lInternalReevaluate(b, now, iWallet, true)
```

Comment:
I haven't checked thoroughly, but I suspect that some of this should have been the attempt activation/retirement calls instead.